types/wlr_primary_selection.c

```diff
@@ -0,0 +1,366 @@
+		wl_resource_get_user_data(resource);
+	wl_signal_emit(&source->events.destroy, source);
+	free(source);
+}
+
+
+void wlr_seat_client_send_primary_selection(
+		struct wlr_seat_client *seat_client) {
+	if (wl_list_empty(&seat_client->primary_selection_devices)) {
+		return;
+	}
+
+	if (seat_client->seat->primary_selection_source) {
+		struct wlr_primary_selection_offer *offer = source_send_offer(
+			seat_client->seat->primary_selection_source, seat_client);
+		if (offer == NULL) {
+			return;
+		}
+
+		struct wl_resource *resource;
+		wl_resource_for_each(resource, &seat_client->primary_selection_devices) {
+			gtk_primary_selection_device_send_selection(resource, offer->resource);
+		}
+	} else {
+		struct wl_resource *resource;
+		wl_resource_for_each(resource, &seat_client->primary_selection_devices) {
+			gtk_primary_selection_device_send_selection(resource, NULL);
+		}
+	}
+}
+
+static void seat_client_primary_selection_source_destroy(
+		struct wl_listener *listener, void *data) {
+	struct wlr_seat *seat =
+		wl_container_of(listener, seat, primary_selection_source_destroy);
+	struct wlr_seat_client *seat_client = seat->keyboard_state.focused_client;
+
+	if (seat_client && seat->keyboard_state.focused_surface) {
+		struct wl_resource *resource;
+		wl_resource_for_each(resource, &seat_client->primary_selection_devices) {
+			gtk_primary_selection_device_send_selection(resource, NULL);
+		}
+	}
+
+	seat->primary_selection_source = NULL;
+
+	wl_signal_emit(&seat->events.primary_selection, seat);
+}
+
+void wlr_seat_set_primary_selection(struct wlr_seat *seat,
+		struct wlr_primary_selection_source *source, uint32_t serial) {
+	if (seat->primary_selection_source &&
+			seat->primary_selection_serial - serial < UINT32_MAX / 2) {
+		return;
+	}
+
+	if (seat->primary_selection_source) {
+		seat->primary_selection_source->cancel(seat->primary_selection_source);
+		seat->primary_selection_source = NULL;
+		wl_list_remove(&seat->primary_selection_source_destroy.link);
+	}
+
+	seat->primary_selection_source = source;
+	seat->primary_selection_serial = serial;
+
+	struct wlr_seat_client *focused_client =
+		seat->keyboard_state.focused_client;
+	if (focused_client) {
+		wlr_seat_client_send_primary_selection(focused_client);
+	}
+
+	wl_signal_emit(&seat->events.primary_selection, seat);
+
+	if (source) {
+		seat->primary_selection_source_destroy.notify =
+			seat_client_primary_selection_source_destroy;
+		wl_signal_add(&source->events.destroy,
+			&seat->primary_selection_source_destroy);
+	}
+}
+
+static void device_handle_set_selection(struct wl_client *client,
+		struct wl_resource *resource, struct wl_resource *source_resource,
+		uint32_t serial) {
+	struct wlr_primary_selection_source *source = NULL;
+	if (source_resource != NULL) {
+		source = wl_resource_get_user_data(source_resource);
+	}
+
+	struct wlr_seat_client *seat_client =
+		wl_resource_get_user_data(resource);
+
+	// TODO: store serial and check against incoming serial here
```

Comment:
How could the serial be validated? Serial validation is currently thought to be a responsibility of rootston. Should we add the ability for the seat itself to validate serials and use this approach in rootston as well?

Reply:
I think #511 is the right way to do this.